blogposts/2022/notebooks-ci.md

```diff
@@ -0,0 +1,93 @@
+</video>
+
+You can read about these in detail in the docs, but the idea is that the pipeline generator primarily determines what gets run over contributions based on:
+
+1. Run types, which are determined by branch naming conventions, tags, and environment variables; and
+2. Diff types, which are determined by what files have been changed in a given branch.
+
+I created notebooks for Run types and Diff types so that Sourcegraph developers can see them in action and in the code. The Notebook for Run types includes both a search that shows Run types in the codebase and a search for RunTypeMatcher, which further shows how each RunType declares the conditions under which it should be used.
+
+Our static CI documentation now contains living documentation–code pulled directly from the current state of the codebase, complete with hover tooltips and jump-to-definition for an IDE-like experience–enabling developers to understand these concepts in theory and in practice. The end result is that our developers will spend fewer hours onboarding and combing through documentation for up-to-date information.
+
+## Notebooks provide a place for on-the-fly, ephemeral documentation
+
+I spend a lot of time explaining code and having code explained to me. One brain can’t hold an entire codebase so we’re always reliant on other developers to know, explain, and document other parts of a codebase. Documentation is rarely sufficient, however, and that often means a call with a shared screen is most effective as one of us walks through code and the other listens.
+
+The trouble with this workflow is that it typically leaves no trace and must happen synchronously, which goes against [Sourcegraph’s async-first aspirations](https://handbook.sourcegraph.com/company-info-and-process/communication/asynchronous-communication/). That’s why I started using Notebooks to build code walkthroughs on the fly.
+
+I found this really useful because it fills a gap I noticed between permanent documentation, which tends to take a while to produce, and synchronous discussions, which can be tough to arrange in a globally remote team. I call it ephemeral documentation.
+
+Let’s walk through a couple of examples. In this Notebook, [Usages of soft failures in Buildkite pipelines](https://sourcegraph.com/notebooks/Tm90ZWJvb2s6NzU1), you can see a brief explanation of soft failures (basically, soft failures don’t cause the entire build to fail) and searches that show soft failures in the Sourcegraph pipeline generator and soft failures in our Buildkite YAML pipelines.
+
+This Notebook came about because a coworker was asking about how we use soft failures in our pipelines, and how they work. Rather than try to list out their usages, I simply wrote a Notebook with some searches that easily captured all usages. From this Notebook, you can use these two searches and see live code that shows exactly what we mean by soft failures.
```

Comment:
```suggestion
This notebook came about because a coworker was asking about how we use soft failures in our pipelines, and how they work. Rather than try to list out their usages, I simply wrote a notebook with some searches that easily captured all usages. From this notebook, you can use these two searches and see live code that shows exactly what we mean by soft failures.
```